src/nni_manager/training_service/reusable/environments/remoteEnvironmentService.ts

```diff
@@ -0,0 +1,309 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+'use strict';
+
+import * as fs from 'fs';
+import * as path from 'path';
+import { Deferred } from 'ts-deferred';
+import * as component from '../../../common/component';
+import { getExperimentId } from '../../../common/experimentStartupInfo';
+import { getLogger, Logger } from '../../../common/log';
+import { EnvironmentInformation, EnvironmentService } from '../environment';
+import { ObservableTimer } from '../../../common/observableTimer';
+import {
+    getExperimentRootDir,
+} from '../../../common/utils';
+import { TrialConfig } from '../../common/trialConfig';
+import { TrialConfigMetadataKey } from '../../common/trialConfigMetadataKey';
+import { execMkdir, validateCodeDir } from '../../common/util';
+import { MachineScheduler } from '../remote/machineScheduler';
+import {
+    ExecutorManager, RemoteMachineMeta,
+} from '../../remote_machine/remoteMachineData';
+import { ShellExecutor } from 'training_service/remote_machine/shellExecutor';
+import { RemoteMachineEnvironmentInformation } from '../remote/remoteConfig';
+
+
+@component.Singleton
+export class RemoteEnvironmentService extends EnvironmentService {
+
+    private readonly initExecutorId = "initConnection";
+    private readonly machineExecutorManagerMap: Map<RemoteMachineMeta, ExecutorManager>;
+    private readonly machineCopyExpCodeDirPromiseMap: Map<RemoteMachineMeta, Promise<void>>;
+    private readonly environmentExecutorManagerMap: Map<string, ExecutorManager>;
```

Comment:
there are 4 maps here, one should be enough.

Reply:
How to use one map? In my point of view, every of these maps has the specific role, and are more logic to use them alone instead of merging them.

Reply:
RemoteMachineMeta and ExecutorManager can be in one data structure. Occupied or not can be tracked by a field in the data structure, or check if executorManager is initialized or not.

Reply:
Generally, we need to maintain a list for all of rmMeta, here we use this machineExecutorManagerMap() instead of a list. If merge RemoteMachineMeta and ExecutorManager  to one data structure, we also need a list to keep the new data structure, I think it's not necessary, the map meets our demands. As for remoteMachineMetaOccupiedMap, the value in this map could be set to false and true dynamically, check executorManager is undefined is not suitable, since it should not be changed to undefined after initialized.